File: background-transcription/android/app/src/main/kotlin/ai/ansight/backgroundtranscription/TranscriptionRunArtifactProvider.kt
package ai.ansight.backgroundtranscription

import ai.ansight.runtime.AndroidArtifactDefinition
import ai.ansight.runtime.AndroidArtifactMetadata
import ai.ansight.runtime.AndroidArtifactProvider
import ai.ansight.runtime.AndroidArtifactProviderDescriptor
import ai.ansight.runtime.AndroidArtifactQueryContext
import ai.ansight.runtime.AndroidArtifactRequest
import ai.ansight.runtime.AndroidArtifactResult

class TranscriptionRunArtifactProvider(
    private val runStore: RunStore,
) : AndroidArtifactProvider {
    override val descriptor = AndroidArtifactProviderDescriptor(
        id = PROVIDER_ID,
        name = "Background transcription runs",
        description = "Exports the immutable JSON result for a terminal background transcription run.",
        category = "transcription",
    )

    override fun query(context: AndroidArtifactQueryContext): List<AndroidArtifactDefinition> {
        val run = runStore.load()
        return listOf(
            AndroidArtifactDefinition(
                id = ARTIFACT_ID,
                name = "Transcript result",
                description = "The saved transcript, run metadata, background-state evidence, and terminal status.",
                kind = "report",
                category = "transcription",
                mimeType = "application/json",
                fileName = "transcription-result.json",
                estimatedSizeBytes = runStore.resultFile.takeIf { it.isFile }?.length(),
                tags = listOf("background", "speech", "transcript"),
                metadata = mapOf(
                    "latestRunId" to run.id,
                    "phase" to run.phase.wireName,
                    "platform" to run.platform,
                ),
            ),
        )
    }

    override fun create(request: AndroidArtifactRequest): AndroidArtifactResult {
        require(request.providerId == PROVIDER_ID && request.artifactId == ARTIFACT_ID) {
            "The transcript artifact request is invalid."
        }
        val requestedRunId = request.arguments["runId"].orEmpty()
        require(requestedRunId.isNotBlank()) { "A runId is required." }

        val run = runStore.load()
        require(run.id == requestedRunId && run.phase.isTerminal) {
            "The requested terminal transcription run is no longer available."
        }
        val bytes = runStore.resultFile.readBytes()
        return AndroidArtifactResult(
            metadata = AndroidArtifactMetadata(
                providerId = request.providerId,
                artifactId = request.artifactId,
                name = "Transcript result ${run.expectedDurationSeconds}s",
                kind = "report",
                mimeType = "application/json",
                fileName = "transcription-${run.expectedDurationSeconds}s-${run.id}.json",
                sizeBytes = bytes.size.toLong(),
                description = "Terminal ${run.platform} background transcription result (${run.phase.wireName}).",
                tags = listOf("background", "speech", "transcript"),
                metadata = mapOf(
                    "runId" to run.id,
                    "platform" to run.platform,
                    "audioSource" to run.audioSource.wireName,
                    "fixtureFileName" to run.fixtureFileName,
                    "backgroundedDuringRun" to run.backgroundedDuringRun.toString(),
                ),
            ),
            bytes = bytes,
        )
    }

    companion object {
        const val PROVIDER_ID = "background-transcription.runs"
        const val ARTIFACT_ID = "transcript-result"
    }
}

File: background-transcription/ios/BackgroundTranscriptionSpike/TranscriptionRunArtifactProvider.swift
import Ansight
import Foundation

struct TranscriptionRunArtifactProvider: AnsightArtifactProvider {
    static let providerId = "background-transcription.runs"
    static let artifactId = "transcript-result"

    let descriptor = AnsightArtifactProviderDescriptor(
        id: providerId,
        name: "Background transcription runs",
        description: "Exports the immutable JSON result for a terminal background transcription run.",
        category: "transcription",
        tags: ["background", "speech", "transcript"]
    )

    func query(context: AnsightArtifactQueryContext) throws -> [AnsightArtifactDefinition] {
        let run = RunStore.shared.load()
        let estimatedSize = (try? FileManager.default.attributesOfItem(atPath: RunStore.shared.resultURL.path)[.size] as? NSNumber)?.int64Value

        return [
            AnsightArtifactDefinition(
                id: Self.artifactId,
                name: "Transcript result",
                description: "The saved transcript, run metadata, background-state evidence, and terminal status.",
                kind: "report",
                category: "transcription",
                content: AnsightArtifactContentDescriptor(
                    supportedMimeTypes: ["application/json"],
                    defaultMimeType: "application/json",
                    suggestedFileName: "transcription-result.json",
                    supportsText: true,
                    supportsBinary: true,
                    sizeKnownBeforeCreation: estimatedSize != nil,
                    estimatedSizeBytes: estimatedSize
                ),
                tags: ["background", "speech", "transcript"],
                metadata: [
                    "latestRunId": run.id,
                    "phase": run.phase.rawValue,
                    "platform": run.platform,
                ]
            ),
        ]
    }

    func create(request: AnsightArtifactRequest) throws -> AnsightArtifactResult {
        guard request.providerId == Self.providerId,
              request.artifactId == Self.artifactId,
              let requestedRunId = request.arguments["runId"],
              !requestedRunId.isEmpty
        else {
            throw ArtifactProviderError.invalidRequest
        }

        let run = RunStore.shared.load()
        guard run.id == requestedRunId, run.phase.isTerminal else {
            throw ArtifactProviderError.runUnavailable
        }

        let path = RunStore.shared.resultURL.path
        let size = (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? NSNumber)?.int64Value
        return AnsightArtifactResult(
            metadata: AnsightArtifactMetadata(
                artifactId: request.artifactId,
                providerId: request.providerId,
                name: "Transcript result \(run.expectedDurationSeconds)s",
                kind: "report",
                mimeType: "application/json",
                fileName: "transcription-\(run.expectedDurationSeconds)s-\(run.id).json",
                description: "Terminal \(run.platform) background transcription result (\(run.phase.rawValue)).",
                sizeBytes: size,
                tags: ["background", "speech", "transcript"],
                metadata: [
                    "runId": run.id,
                    "platform": run.platform,
                    "audioSource": run.audioSource ?? TranscriptionSource.injectedFixture.rawValue,
                    "fixtureFileName": run.fixtureFileName,
                    "backgroundedDuringRun": String(run.backgroundedDuringRun),
                ]
            ),
            payload: .fromFile(path)
        )
    }
}

private enum ArtifactProviderError: LocalizedError {
    case invalidRequest
    case runUnavailable

    var errorDescription: String? {
        switch self {
        case .invalidRequest:
            return "The transcript artifact request is invalid."
        case .runUnavailable:
            return "The requested terminal transcription run is no longer available."
        }
    }
}
